### ISSUE TITLE
Selective eager auto materialization

### ISSUE THREAD

**Issue by pablo-statsig** [dagster-io/dagster#15549]
### What's the use case?

I want to be able to have an asset that auto materializes eagerly but only based on some parents updating.

Basically if I have asset A that depends on B, C and B updates every 5 minutes and C updates every day I would like to make it so that A will materialize when B, C is ready but will only re-materialzie if C is updated not if B is updated.  

### Ideas of implementation

Basically, It would be nice to be able to categorize upstream assets into lazy or eager so that the downstream asset can behave one way or another depending on  what assets have been updated.

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by sryza** [comment]
Thanks for filing this @pablo-statsig. Tracking this along with a set of other requests for customizing auto-materialize policies.

**Comment by OwenKephart** [comment]
Hi @pablo-statsig ! In this particular case I think you could get away with a more general solution. In particular, we've recently added a new `AutoMaterializeRule.skip_on_not_all_parents_updated()` rule, which will cause an asset to only be materialized once _all_ of its parents have updated. That looks like:

```
from dagster import AutoMaterializePolicy, AutoMaterializeRule

my_policy = AutoMaterializePolicy.eager().with_rules(
    AutoMaterializeRule.skip_on_not_all_parents_updated(),
)
```

This would result in the behavior:

1. `B` and `C` are materialized, so `A` is materialized
2. 5 minutes later, `B` is materialized again, but `C` hasn't been updated since `A` ran, so `A` is not materialized.
3. the above process continues every 5 minutes for the next 24 hours...
4. now, `C` is materialized, so both of `A`'s parents have been materialized since its last materialization, and it is materialized.

I believe the above behavior is what you're looking for, is that right?

That's not to say that there's no place for selectively ignoring updates to specific upstream assets, but in the example you proposed I think the simpler solution is better. Do you have thoughts on if you have other situations in which you'd need the original requested behavior?

**Comment by pablo-statsig** [comment]
yeah that seems to fit my use case!

**Comment by garethbrickman** [comment]
Worth checking/commenting in this RFC for the forthcoming Declarative Automation feature: https://github.com/dagster-io/dagster/discussions/22811
